### PR TITLE
Support changes to detached nodes in OptionalField

### PR DESCRIPTION
**Pull request by Abe27342**
## Description

Adds support for making changes to detached nodes in optional field. This includes a general change in the model used to think about optional field.

This PR doesn't yet fully fix problems with optional field obeying desired rebaser axioms, but it improves the correctness situation such that more axiomatic tests and a reasonable number of fuzz tests now pass.

### CODE REVIEW

File: experimental/dds/tree2/src/test/shared-tree/editing.spec.ts
Comment (Abe27342) [inline experimental/dds/tree2/src/test/shared-tree/editing.spec.ts:814]: I left some test cases around in this file for sequence field which were interesting for optional field (in that at some point I had an implementation which failed the test case, and I wondered how sequence field got it right).

I'm ok removing them or shifting to a different PR if we care.

File: experimental/dds/tree2/src/test/shared-tree/editing.spec.ts
Comment (Abe27342) [inline experimental/dds/tree2/src/test/shared-tree/editing.spec.ts:2577]: this test is still skipped as it fails for sequence field, but it was my understanding that we're likely to support the correct merge semantics here before we support forking of the undo/redo stack, and the same scenario was possible to model.

File: experimental/dds/tree2/src/test/feature-libraries/optional-field/optionalChangeRebaser.spec.ts
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/test/feature-libraries/optional-field/optionalChangeRebaser.spec.ts:146]: Why is this ` ?? 0) + 1` stuff needed?

File: experimental/dds/tree2/src/test/feature-libraries/optional-field/optionalChangeRebaser.spec.ts
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/test/feature-libraries/optional-field/optionalChangeRebaser.spec.ts:221]: Looks like this part of the logic could just reuse `getSequentialEdits`, or, even better, take in the array as parameter instead of recomputing it here (in which case you would iterate it backwards instead of reversing it).

File: experimental/dds/tree2/src/test/feature-libraries/optional-field/optionalChangeRebaser.spec.ts
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/test/feature-libraries/optional-field/optionalChangeRebaser.spec.ts:326]: It might get confusing if one changes `newContents` without changing `description` or vice-versa.
```suggestion
				description: `Set${newContents}`,
```

File: experimental/dds/tree2/src/test/feature-libraries/optional-field/optionalChangeRebaser.spec.ts
Comment (Abe27342) [reply experimental/dds/tree2/src/test/feature-libraries/optional-field/optionalChangeRebaser.spec.ts:146]: It could just be `const idAllocator = failIdAllocator;` now: at one point while working on this change the rebase logic for optional field did sometimes use the id allocator passed to it, so it was necessary to pass a legitimate one.

If you take for granted that optional field needs one here, I think the `?? 0) + 1` is reasonable: when rebasing a child-only change over another child-only change, neither change has any local ids, and getMaxId returns undefined there. +1 guarantees no reuse of ids. Admittedly I wasn't particularly careful about whether the id allocator returns the value pre-increment or post-increment, so maybe it's not necessary. but not allocating an id isn't a big deal :)

File: experimental/dds/tree2/src/test/feature-libraries/optional-field/optionalField.spec.ts
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/test/feature-libraries/optional-field/optionalField.spec.ts:109]: I would have expected `["self", { localId: brand(42) }, "nodeTargeting"],` and a new `reservedDetachId`

File: experimental/dds/tree2/src/feature-libraries/optional-field/optionalFieldChangeTypes.ts
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/feature-libraries/optional-field/optionalFieldChangeTypes.ts:61]:
```suggestion
	 * In such cases, this changeset should not include a move with source `foo`, since `foo` is empty.
```

File: experimental/dds/tree2/src/test/feature-libraries/optional-field/optionalField.spec.ts
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/test/feature-libraries/optional-field/optionalField.spec.ts:391]: Adding a  move from { localId: brand(0), revision: taggedBaseChange.revision } to { localId: brand(1) } doesn't seem like a viable solution here because { localId: brand(1) } should contain the node inserted by baseChange and removed by the rebased change.

File: experimental/dds/tree2/src/test/feature-libraries/optional-field/optionalField.spec.ts
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/test/feature-libraries/optional-field/optionalField.spec.ts:396]: Maybe mention that `reservedDetachId` shound not be set here.

File: experimental/dds/tree2/src/test/shared-tree/editing.spec.ts
Comment (yann-achard-MS) [reply experimental/dds/tree2/src/test/shared-tree/editing.spec.ts:814]: Happy to keep them as part of this PR.

Comment (yann-achard-MS) [reply experimental/dds/tree2/src/test/shared-tree/editing.spec.ts:814]: Maybe rename to this though:
```suggestion
		it("delete, undo, childchange rebased over childchange", () => {
```

File: experimental/dds/tree2/src/test/shared-tree/editing.spec.ts
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/test/shared-tree/editing.spec.ts:849]:
```suggestion
		it("childchange rebase over delete, undo, childchange", () => {
```

File: experimental/dds/tree2/src/test/shared-tree/editing.spec.ts
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/test/shared-tree/editing.spec.ts:1873]: Seems correct to me, but hard to follow. May I interest you in a different approach?
```suggestion
				tree1.merge(tree2, false);
				tree1.merge(tree3, false);
				tree2.rebaseOnto(tree1);
				tree3.rebaseOnto(tree1);
```
The first two lines define the edit sequencing, while the last two just represent the participants learning of that ordering.

File: experimental/dds/tree2/src/test/shared-tree/editing.spec.ts
Comment (yann-achard-MS) [reply experimental/dds/tree2/src/test/shared-tree/editing.spec.ts:2577]: Thanks, it's nice that the test doesn't need the undo/redo stack forking.
I don't know whether we'll want to supporting editing already removed content in this way anytime soon. It seems like a somewhat advanced use case.

File: experimental/dds/tree2/src/feature-libraries/optional-field/optionalFieldChangeTypes.ts
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/feature-libraries/optional-field/optionalFieldChangeTypes.ts:40]: Consider clarifying that we're talking about the source of the move here.

File: experimental/dds/tree2/src/feature-libraries/optional-field/optionalFieldChangeFormat.ts
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/feature-libraries/optional-field/optionalFieldChangeFormat.ts:20]: Could we make this a tuple for some additional savings? Doesn't have to happen in this PR though.

File: experimental/dds/tree2/src/test/feature-libraries/optional-field/optionalChangeRebaser.spec.ts
Comment (yann-achard-MS) [reply experimental/dds/tree2/src/test/feature-libraries/optional-field/optionalChangeRebaser.spec.ts:146]: I'm confused: if neither changes have IDs then how could there be any reuse of IDs?
Note that `idAllocatorFromMaxId`...
- accepts undefined (in which case the next ID it generates will be 0). 
- will next generate N+1 when seeded with N.

File: experimental/dds/tree2/src/feature-libraries/optional-field/optionalField.ts
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/feature-libraries/optional-field/optionalField.ts:37]: Consider renaming to `IRegisterMap` or `IRegistry` since it is now used for more than just child changes.
Same for `ChildChangeMap`

File: experimental/dds/tree2/src/feature-libraries/optional-field/optionalField.ts
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/feature-libraries/optional-field/optionalField.ts:176]: Consider refactoring into
```suggestion
			isInputContextEmpty ||= tryInferInputContext(change);
```
where `tryInferInputContext` would return `undefined` when `change.moves.length > 0 || change.reservedDetachId !== undefined` is false, and `change.reservedDetachId !== undefined` otherwise.
The main benefit I see for it is that making `tryInferInputContext` a separate function with a clear contract will makes it easier to assess its correctness (now and in the future).

Correctness-wise, it seems like a leap to infer that the presence of move entries implies that at least one of those moves affect the "self" register. There's at least nothing in the `OptionalChangeset` data model that guarantees this is the case. So some comment explainling the reasoning for this leap (or an actual run-time check) would be good.

File: experimental/dds/tree2/src/feature-libraries/optional-field/optionalField.ts
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/feature-libraries/optional-field/optionalField.ts:189]: Since inverses never produce builds it shoud be sufficient to just do:
```suggestion
							revision: id.revision ?? revision,
```
I'm not really concerned about the perf, I mostly want to keep the use of `revisionMetadata` to a minimum so we can keep track of when and why it is needed.

File: experimental/dds/tree2/src/feature-libraries/optional-field/optionalField.ts
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/feature-libraries/optional-field/optionalField.ts:187]: The nature of sandwich rebasing is such that we may compose changes that attempt to build the same node multiple times. These would then end up a duplicates in the builds list.
Even without considering sandwich rebasing, it seems good to have a policy for this.
Either maintain a set to weed out the dupes, or add a to call it out.

File: experimental/dds/tree2/src/feature-libraries/optional-field/optionalField.ts
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/feature-libraries/optional-field/optionalField.ts:202]: This is another case where we'll never actually need the metadata. It's convenient to just use this helper though, so it's fine to keep it that way for now and we'll try to remove the use of all metadata later on.

File: experimental/dds/tree2/src/feature-libraries/optional-field/optionalField.ts
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/feature-libraries/optional-field/optionalField.ts:207]: Consider de-duplicating all the evaluations for `withIntention(src)` and `withIntention(dst)`.

File: experimental/dds/tree2/src/feature-libraries/optional-field/optionalField.ts
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/feature-libraries/optional-field/optionalField.ts:224]: It's hard to look at this and grok what it means for all 6 combinations.

I think one little thing that would have helped me woud be to model the case where there is no `originalSrc` with an undefined `currentTarget` (instead of setting it to `cellTargeting`, which is arbitrary).

Then perhaps this expression becomes easier to read:
```suggestion
					target === "cellTargeting" && currentTarget === "cellTargeting"
						? "cellTargeting"
						: "nodeTargeting",
```

File: experimental/dds/tree2/src/feature-libraries/optional-field/optionalField.ts
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/feature-libraries/optional-field/optionalField.ts:233]: Consider de-duplicating the evaluations for `withIntention(dst)`.

File: experimental/dds/tree2/src/feature-libraries/optional-field/optionalField.ts
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/feature-libraries/optional-field/optionalField.ts:291]:
```suggestion
		const invertedMoves: typeof change.moves = [];
```

File: experimental/dds/tree2/src/feature-libraries/optional-field/optionalField.ts
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/feature-libraries/optional-field/optionalField.ts:293]:
```suggestion
			// TODO:AB#6298: This assert can legitimately fail for transactions, meaning we have little test coverage there.
```

File: experimental/dds/tree2/src/feature-libraries/optional-field/optionalField.ts
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/feature-libraries/optional-field/optionalField.ts:296]: I'm having trouble figuring out what "single-register attached sets" means. Can this be rephrased in more pedestrian terms? I must say I'm struggling to come up with a way of putting it.

Comment (yann-achard-MS) [reply experimental/dds/tree2/src/feature-libraries/optional-field/optionalField.ts:296]: Perhaps it's easier to say what isn't supported: "changes that transfer nodes between non-self registers".

File: experimental/dds/tree2/src/feature-libraries/optional-field/optionalField.ts
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/feature-libraries/optional-field/optionalField.ts:298]: Sorry, this is really nitpicky but I keep reading this wrong (it doesn't help that it lines up with the assert above).

```suggestion
			if (src !== "self" && dst === "self") {
```

File: experimental/dds/tree2/src/feature-libraries/optional-field/optionalField.ts
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/feature-libraries/optional-field/optionalField.ts:299]: This might lead to a situation where we put a node in a register it never came from, and that register may not be empty.
Repro:
- E1: Replace A with B
- E2: undo E1 (concurrently to E3)
- E3: undo E1 (concurrently to E2)
I'm ok with just leaving a comment for now if you want to merge before addressing this.

File: experimental/dds/tree2/src/feature-libraries/optional-field/optionalField.ts
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/feature-libraries/optional-field/optionalField.ts:367]: The code in this if block is assuming that the only source register that is being moved from with cell targeting is "self", right?
If so, an assert or comment would be nice.

File: experimental/dds/tree2/src/feature-libraries/optional-field/optionalField.ts
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/feature-libraries/optional-field/optionalField.ts:387]: Nice call-out :-)
Funny coincidence: I added the same exact logic to sequence moves yesterday.

File: experimental/dds/tree2/src/feature-libraries/optional-field/optionalField.ts
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/feature-libraries/optional-field/optionalField.ts:426]: I'm curious what motivated you drop the new build when rebasing over a build for the same ID.
Do we even expect that to happen ever? I think we don't right now, because there's no way for a branch to get the same change twice.
Once we allow merging other branches as part of transanctions then perhaps that will start occuring. When that's the case I'm still expecting that we may not want to filter.
Happy to chat about this one.

File: experimental/dds/tree2/src/feature-libraries/optional-field/optionalField.ts
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/feature-libraries/optional-field/optionalField.ts:356]: I must say I'm having trouble understanding what `renamedDsts` represents. Which pair of contexts is this spanning?
It's going away fairly soon though, so I'm okay with it.

File: experimental/dds/tree2/src/feature-libraries/optional-field/optionalField.ts
Comment (yann-achard-MS) [reply experimental/dds/tree2/src/feature-libraries/optional-field/optionalField.ts:187]: All the build stuff is about to be pulled out of FieldKinds so this doesn't matter very much.

File: experimental/dds/tree2/src/test/feature-libraries/optional-field/optionalChangeRebaser.spec.ts
Comment (Abe27342) [reply experimental/dds/tree2/src/test/feature-libraries/optional-field/optionalChangeRebaser.spec.ts:146]: I'm sure `idAllocatorFromMaxId(getMaxId(change, base.change))` is sufficient then. I wasn't aware of either property of `idAllocatorFromMaxId` when implementing and didn't want to bother checking for test code.

File: experimental/dds/tree2/src/test/feature-libraries/optional-field/optionalField.spec.ts
Comment (Abe27342) [reply experimental/dds/tree2/src/test/feature-libraries/optional-field/optionalField.spec.ts:396]: I think it depends on which strategy we use to resolve the above TODO. I'm leaning toward `expected`'s child changes referring to `{ localId: 0 }` and keeping reservedDetachId, but haven't thought super deep about it since writing this comment. I'll leave a note though. That should make it clear enough the expectation here isn't desired long-term :)

File: experimental/dds/tree2/src/feature-libraries/optional-field/optionalFieldChangeFormat.ts
Comment (Abe27342) [reply experimental/dds/tree2/src/feature-libraries/optional-field/optionalFieldChangeFormat.ts:20]: might as well. There's still some more low-hanging fruit in this realm. I think encoding `self` as `null` and ChangeAtomIds without revision as just their local id is another quick win. I'll do that in a followup though.

Comment (Abe27342) [reply experimental/dds/tree2/src/feature-libraries/optional-field/optionalFieldChangeFormat.ts:20]: https://dev.azure.com/fluidframework/internal/_workitems/edit/6315

File: experimental/dds/tree2/src/feature-libraries/optional-field/optionalField.ts
Comment (Abe27342) [reply experimental/dds/tree2/src/feature-libraries/optional-field/optionalField.ts:176]: Thanks for the suggestion and prompting me to look into this a bit more--it bugged me too. Turns out that not making this logical leap was only blocked on generating `reservedDetachId`s for inverses (e.g. inverting `OptionalField.clear(wasEmpty: false, { localId: 0 })`), which is something we should be doing anyway.

Next update, the impl here should follow much more naturally from documented invariants on optional changesets.

File: experimental/dds/tree2/src/feature-libraries/optional-field/optionalField.ts
Comment (Abe27342) [reply experimental/dds/tree2/src/feature-libraries/optional-field/optionalField.ts:224]: I don't think that expression works--when processing a move that has no `originalSrc` (i.e. the move's pre-existing source *is* its original source), we'll have `currentTarget` undefined and `target` is `cellTargeting`, but end up with `nodeTargeting`. So e.g. `compose({ moves: [["self", { localId: 1 }, "cellTargeting"]] })` would suddenly become `nodeTargeting`.

That said, I see your point about initializing to undefined making the decision point more uniform. I'll do that.

File: experimental/dds/tree2/src/feature-libraries/optional-field/optionalField.ts
Comment (Abe27342) [reply experimental/dds/tree2/src/feature-libraries/optional-field/optionalField.ts:356]: Added this explanation and moved this to be more colocated with where it's used:

```
		// Maps the content occupying a given register id in the output context of `change` to the register id
		// that content occupies in the output context of the rebased change.
		// This is necessary since child changes to nodes are keyed on the register they occupy in the output
		// context of the changeset, and `change` might make child changes to content which it doesn't move, but
		// `over` does.
```

File: experimental/dds/tree2/src/feature-libraries/optional-field/optionalField.ts
Comment (Abe27342) [reply experimental/dds/tree2/src/feature-libraries/optional-field/optionalField.ts:426]: Hah. You're correct, thanks for checking me here. There were two things going on:

1. When I added this code, I wasn't properly tagging basically any `RegisterId`s with appropriate revisions. So I'd end up with random collisions of build ids for things that don't really make sense with modular changeset's design. Interestingly, pedagogically speaking this experience made me a bit more comfortable with the interplay between change rebasers imbuing changesets with revisions and modular change family: I think that bit of the puzzle is a huge part of what I was missing while toiling away on optional field before.
2. builds weren't idempotent when I wrote this, so collision actually resulted in failure.

It's totally not necessary anymore, and is removed.

File: experimental/dds/tree2/src/feature-libraries/optional-field/optionalField.ts
Comment (Abe27342) [inline experimental/dds/tree2/src/feature-libraries/optional-field/optionalField.ts:471]: @yann-achard-MS -- meant to ask at some point, but what are the requirements on `amendRebase`? I have a vague idea it's to support cross-field moves of content, which makes me think maybe this implementation is ok. not sure though :)

Comment (Abe27342) [reply experimental/dds/tree2/src/feature-libraries/optional-field/optionalField.ts:471]: nvm, I see it's superceded by `prune` on latest main anyway.

Comment (Abe27342) [reply experimental/dds/tree2/src/feature-libraries/optional-field/optionalField.ts]: would like a close pair of eyes on the prune implementation. One top-level fuzz test failed on latest, but it seems likely unrelated to optional field based on failure JSON. If there isn't work filed for that, I can file it now.

File: experimental/dds/tree2/src/feature-libraries/optional-field/optionalField.ts
Comment (yann-achard-MS) [reply experimental/dds/tree2/src/feature-libraries/optional-field/optionalField.ts]: Ah good point, thanks.

File: experimental/dds/tree2/src/feature-libraries/optional-field/optionalField.ts
Comment (yann-achard-MS) [reply experimental/dds/tree2/src/feature-libraries/optional-field/optionalField.ts]: prune implementation looks good :)

**Review by yann-achard-MS — APPROVED** [review]

File: experimental/dds/tree2/src/test/shared-tree/editing.spec.ts
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/test/shared-tree/editing.spec.ts:826]: Rename the test to "delete, undo, childchange rebased over childchange"

**Pull request merged**
